csv/generic/src/main/scala/fs2/data/csv/generic/MapShapedCsvRowDecoder.scala

```diff
@@ -27,80 +27,142 @@ trait MapShapedCsvRowDecoder[Repr] extends CsvRowDecoder[Repr, String]
+                                     EmbedTail <: HList](implicit
       witness: Witness.Aux[Key],
       Head: CellDecoder[Head],
-      ev: <:<[Anno, Option[CsvName]],
-      Tail: Lazy[WithDefaults[Wrapped, Tail, DefaultTail, AnnoTail]])
+      ev: <:<[Name, Option[CsvName]],
+      Tail: Lazy[WithDefaults[Wrapped, Tail, DefaultTail, NamesTail, EmbedTail]])
       : WithDefaults[Wrapped,
                      FieldType[Key, Option[Head]] :: Tail,
                      Option[Option[Head]] :: DefaultTail,
-                     Anno :: AnnoTail] =
-    new WithDefaults[Wrapped,
+                     Name :: NamesTail,
+                     None.type :: EmbedTail] =
+    (row: CsvRow[String],
+     default: Option[Option[Head]] :: DefaultTail,
+     names: Name :: NamesTail,
+     embeds: None.type :: EmbedTail) => {
+      val head = row(names.head.fold(witness.value.name)(_.name)) match {
+        case Some(head) if head.nonEmpty => Head(head).map(Some(_))
+        case _                           => Right(default.head.flatten)
+      }
+      for {
+        head <- head
+        tail <- Tail.value.fromWithDefault(row, default.tail, names.tail, embeds.tail)
+      } yield field[Key](head) :: tail
+    }
+
+  implicit def optionHconsEmbedRowDecoder[Wrapped,
+                                          Key <: Symbol,
+                                          Head,
+                                          Tail <: HList,
+                                          DefaultTail <: HList,
+                                          NamesTail <: HList,
+                                          EmbedTail <: HList](implicit
+      witness: Witness.Aux[Key],
+      Head: CsvRowDecoder[Option[Head], String],
+      Tail: Lazy[WithDefaults[Wrapped, Tail, DefaultTail, NamesTail, EmbedTail]])
+      : WithDefaults[Wrapped,
                      FieldType[Key, Option[Head]] :: Tail,
                      Option[Option[Head]] :: DefaultTail,
-                     Anno :: AnnoTail] {
-      def fromWithDefault(row: CsvRow[String],
-                          default: Option[Option[Head]] :: DefaultTail,
-                          anno: Anno :: AnnoTail): DecoderResult[FieldType[Key, Option[Head]] :: Tail] = {
-        val head = row(anno.head.fold(witness.value.name)(_.name)) match {
-          case Some(head) if head.nonEmpty => Head(head).map(Some(_))
-          case _                           => Right(default.head.flatten)
+                     None.type :: NamesTail,
+                     Some[CsvEmbed] :: EmbedTail] =
+    (row: CsvRow[String],
+     default: Option[Option[Head]] :: DefaultTail,
+     names: None.type :: NamesTail,
+     embeds: Some[CsvEmbed] :: EmbedTail) => {
+      for {
+        head <- (Head(row), default.head) match {
+          case (r @ Right(_), _)                                    => r
+          case (Left(_: DecoderError.ColumnMissing), Some(default)) => Right(default)
```

Comment:
This is a somewhat tricky case as it could be that some of the fields of the embedded value are present, some are invalid. In this edge case, it depends on the order of fields whether the default is taken or the error of an invalid field is bubbled up. The only way to mitigate this would be to introduce error-accumulating parsing (which would indeed be an interesting addition), but as the exact combination of 
* embedding 
* invalid fields for the embedded value 
* missing fields for the embedded value and
* a default for the embedded field 

should be extremely rare in the wild, I believe this is fine.